// gyazo-api/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _request = require('request');

var _request2 = _interopRequireDefault(_request);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Gyazo = function () {
  function Gyazo() {
    var accessToken = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

    _classCallCheck(this, Gyazo);

    this.accessToken = accessToken;
  }

  _createClass(Gyazo, [{
    key: 'upload',
    value: function upload(image) {
      var _this = this;

      var params = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      return new Promise(function (resolve, reject) {
        if (!image) throw new Error('image is undefined');
        if (typeof image === 'string') image = _fs2.default.createReadStream(image);
        var url = 'https://upload.gyazo.com/api/upload';
        var req = _request2.default.post({
          url: url
        }, function (err, res, body) {
          if (err) return reject(err);
          if (res.statusCode !== 200) return reject(res.body);
          resolve({
            response: res,
            data: JSON.parse(body)
          });
        });
        var form = req.form();
        form.append('imagedata', image);
        form.append('access_token', _this.accessToken);
        for (var k in params) {
          form.append(k, params[k]);
        }
      });
    }
  }, {
    key: 'list',
    value: function list() {
      var _this2 = this;

      var query = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

      return new Promise(function (resolve, reject) {
        query.access_token = _this2.accessToken;
        var url = 'https://api.gyazo.com/api/images';
        _request2.default.get({
          url: url,
          qs: query
        }, function (err, res, body) {
          if (err) return reject(err);
          if (res.statusCode !== 200) return reject(res.body);
          resolve({
            response: res,
            data: JSON.parse(res.body)
          });
        });
      });
    }
  }, {
    key: 'delete',
    value: function _delete(imageId) {
      var _this3 = this;

      return new Promise(function (resolve, reject) {
        if (!imageId) throw new Error('imageId is undefined');
        var url = 'https://api.gyazo.com/api/images/' + imageId;
        _request2.default.del({
          url: url,
          qs: {
            access_token: _this3.accessToken
          }
        }, function (err, res, body) {
          if (err) return reject(err);
          if (res.statusCode !== 200) return reject(res.body);
          resolve({
            response: res,
            data: JSON.parse(res.body)
          });
        });
      });
    }
  }]);

  return Gyazo;
}();

exports.default = Gyazo;
